Add support for the `allow-downloads` iframe sandbox attribute

robert-jenner opened the pull request:
#### f8f6da20d5073dc5050556c3f4e1c6f8007b2b13
<pre>
Add support for the `allow-downloads` iframe sandbox attribute
<a href="https://bugs.webkit.org/show_bug.cgi?id=259155">https://bugs.webkit.org/show_bug.cgi?id=259155</a>
rdar://109414664

Reviewed by Brent Fulgham.

Add support for the `allow-downloads` iframe sandbox attribute so that downloads
are now forbidden by default in sandboxed iframes and this attribute is required
to allow them. This aligns our behavior with Chrome and Firefox.

* LayoutTests/http/tests/download/sandboxed-iframe-download-allowed-expected.txt: Added.
* LayoutTests/http/tests/download/sandboxed-iframe-download-allowed-in-popup-expected.txt: Added.
* LayoutTests/http/tests/download/sandboxed-iframe-download-allowed-in-popup-noopener-expected.txt: Added.
* LayoutTests/http/tests/download/sandboxed-iframe-download-allowed-in-popup-noopener.html: Added.
* LayoutTests/http/tests/download/sandboxed-iframe-download-allowed-in-popup.html: Added.
* LayoutTests/http/tests/download/sandboxed-iframe-download-allowed.html: Added.
* LayoutTests/http/tests/download/sandboxed-iframe-download-not-allowed-expected.txt: Added.
* LayoutTests/http/tests/download/sandboxed-iframe-download-not-allowed.html: Added.
Add layout test coverage.

* LayoutTests/imported/w3c/web-platform-tests/html/semantics/embedded-content/the-iframe-element/iframe_sandbox_anchor_download_block_downloads.tentative-expected.txt:
Rebaseline due to more PASS lines. I investigated the remaining failure and confirmed that no download is actually occurring for any of the subtests, as expected.
I see our code cancelling the NetworkDataTask. However, the test doesn&apos;t seem to detect the stream closing. It may be an issue as CFNetwork level.

* Source/WebCore/dom/SecurityContext.cpp:
(WebCore::SecurityContext::isSupportedSandboxPolicy):
(WebCore::SecurityContext::parseSandboxPolicy):
* Source/WebCore/dom/SecurityContext.h:
* Source/WebCore/loader/DocumentLoader.cpp:
(WebCore::DocumentLoader::continueAfterContentPolicy):
* Source/WebCore/loader/PolicyChecker.cpp:
(WebCore::FrameLoader::PolicyChecker::checkNavigationPolicy):
(WebCore::FrameLoader::PolicyChecker::checkNewWindowPolicy):

* Tools/WebKitTestRunner/TestController.cpp:
(WTR::TestController::decidePolicyForNavigationResponse):
Update WebKitTestRunner to trigger a download when a resource response contains
`Content-Disposition: attachment`, similarly to what a browser would do and what
WPT tests expect. By default, we would only trigger a download if the mime type
couldn&apos;t be displayed. However, iframe_sandbox_anchor_download_block_downloads.tentative.html
WPT test uses both a `Content-Disposition: attachment` HTTP header and a displayable
MIME type and rightly expects a download.

Originally-landed-as: 265870.84@safari-7616-branch (9d46b7777bca). rdar://116425379
Canonical link: <a href="https://commits.webkit.org/269788@main">https://commits.webkit.org/269788@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0ea7974cca0a03b76e1a4e8d151adf8bf4efd2ef

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/23553 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/1670 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/24675 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/25714 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/21746 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/23822 "Passed tests") | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/49/builds/3278 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/24093 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/22334 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/23795 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/1219 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/20380 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/26310 "Built successfully") | 
| | [❌ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/1017 "Unexpected infrastructure issue, retrying build") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/21279 "Passed tests") | [⏳ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/GTK-WK2-Tests-EWS "Waiting to run tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/21489 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/21544 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/25321 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/992 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/18707 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/964 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/21053 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/5632 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/1404 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/1271 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Tools/WebKitTestRunner/TestController.h:
```diff
@@ -756,6 +757,9 @@ class TestController {
     bool m_shouldLogDownloadSize { false };
     bool m_shouldLogDownloadExpectedSize { false };
+    size_t m_downloadIndex { 0 };
+    bool m_shouldLogDownloadSize { false };
```
This is already present at line 758. The build error is about this data member being defined twice. You need to drop this one.

cdumez committed:
Add support for the `allow-downloads` iframe sandbox attribute
https://bugs.webkit.org/show_bug.cgi?id=259155
rdar://109414664

Reviewed by Brent Fulgham.

Add support for the `allow-downloads` iframe sandbox attribute so that downloads
are now forbidden by default in sandboxed iframes and this attribute is required
to allow them. This aligns our behavior with Chrome and Firefox.

* LayoutTests/http/tests/download/sandboxed-iframe-download-allowed-expected.txt: Added.
* LayoutTests/http/tests/download/sandboxed-iframe-download-allowed-in-popup-expected.txt: Added.
* LayoutTests/http/tests/download/sandboxed-iframe-download-allowed-in-popup-noopener-expected.txt: Added.
* LayoutTests/http/tests/download/sandboxed-iframe-download-allowed-in-popup-noopener.html: Added.
* LayoutTests/http/tests/download/sandboxed-iframe-download-allowed-in-popup.html: Added.
* LayoutTests/http/tests/download/sandboxed-iframe-download-allowed.html: Added.
* LayoutTests/http/tests/download/sandboxed-iframe-download-not-allowed-expected.txt: Added.
* LayoutTests/http/tests/download/sandboxed-iframe-download-not-allowed.html: Added.
Add layout test coverage.

* LayoutTests/imported/w3c/web-platform-tests/html/semantics/embedded-content/the-iframe-element/iframe_sandbox_anchor_download_block_downloads.tentative-expected.txt:
Rebaseline due to more PASS lines. I investigated the remaining failure and confirmed that no download is actually occurring for any of the subtests, as expected.
I see our code cancelling the NetworkDataTask. However, the test doesn't seem to detect the stream closing. It may be an issue as CFNetwork level.

* Source/WebCore/dom/SecurityContext.cpp:
(WebCore::SecurityContext::isSupportedSandboxPolicy):
(WebCore::SecurityContext::parseSandboxPolicy):
* Source/WebCore/dom/SecurityContext.h:
* Source/WebCore/loader/DocumentLoader.cpp:
(WebCore::DocumentLoader::continueAfterContentPolicy):
* Source/WebCore/loader/PolicyChecker.cpp:
(WebCore::FrameLoader::PolicyChecker::checkNavigationPolicy):
(WebCore::FrameLoader::PolicyChecker::checkNewWindowPolicy):

* Tools/WebKitTestRunner/TestController.cpp:
(WTR::TestController::decidePolicyForNavigationResponse):
Update WebKitTestRunner to trigger a download when a resource response contains
`Content-Disposition: attachment`, similarly to what a browser would do and what
WPT tests expect. By default, we would only trigger a download if the mime type
couldn't be displayed. However, iframe_sandbox_anchor_download_block_downloads.tentative.html
WPT test uses both a `Content-Disposition: attachment` HTTP header and a displayable
MIME type and rightly expects a download.

Originally-landed-as: 265870.84@safari-7616-branch (9d46b7777bca). rdar://116425379
Canonical link: https://commits.webkit.org/269788@main
